src/cockpit/channels/dbus.py

```diff
@@ -205,17 +244,44 @@ def do_open(self, options):
             if err.errno != errno.EBUSY:
                 raise
 
-        self.ready()
-
-    def add_match(self, rule, handler):
+        if self.name is not None:
+            async def get_ready():
+                async with self.watch_processing_lock:
+                    await self.setup_name_owner_tracking()
+                    if self.owner:
+                        self.ready(unique_name=self.owner)
+                    else:
+                        self.close(problem="not-found")
+            task = asyncio.create_task(get_ready())
+            self.tasks.add(task)
+            task.add_done_callback(self.tasks.discard)
+        else:
+            self.ready()
+
+    def add_signal_handler(self, rule, handler):
+        r = rule.copy()
+        r['type'] = 'signal'
+        if 'sender' not in r and self.name:
+            r['sender'] = self.name
+        if r.get('path_namespace') == "/":
+            del r['path_namespace']
+        def filter_owner(message):
+            if r['sender'] == self.name:
+                if self.owner and self.owner != message.get_sender():
```

Comment:
Is this some kind of attempt to protect against races where we receive signals from a new name owner before we're aware that the name has a new owner?  Is that even possible?  We'll have seen the `NameOwnerChanged` signal first...

I'd lean towards thinking that this shouldn't be necessary, but the fact that you've included it sort of indicates that probably it *is* necessary for some reason.  This definitely needs at least a comment.

Reply:
> Is this some kind of attempt to protect against races

No, this is to filter out signals that come from other senders.  Filtering by a well-known name doesn't seem to work, and we would receive generic signals like PropertiesChanged or InterfacesAdded from the wrong sender.

I think dbus-broker should be able to filter by well-known sender names, but sd-bus probably can not.

Reply:
This is strange.  I also had a memory that filtering by non-unique names didn't work, so I checked it.

The D-Bus spec certainly thinks that you can filter by well-known name.  It includes a couple of examples of doing exactly that.

Reply:
Yes, calling AddMatch on the bus daemon with a well-known name should work.  The daemon knows the well-known name for all unique names when it processes the match rules.

But sd-bus (or glib) also need to process the match rules when a signal comes in, and they only know the unique name of the sender.

Our bridge will call AddMatch multiple times on a single connection.  For example, the Networking page will call bus.add_match for sender=org.freedesktop.NetworkManager and sender=org.freedesktop.systemd (both with interface=org.freedesktop.DBus.Properties).

The handler given to the bus.add_match call for NetworkManager will then see the signals for systemd, because they will both come from the dbus daemon, bu sd-bus can't tell what the well-known name of the signal is.

dbus.py knows, so we can complete the filtering there.

(That's my theory, at least.)

Reply:
I wonder if there is a sd-bus bug here.  Would it do a better job if (for example) we used its built-in name-tracking code?  Might be worth a report upstream.